### ISSUE TITLE
Kaggle Experiments

### ISSUE THREAD

**Issue by benitoThree** [ThirdAILabs/Universe#242]
Parent issue: #225 

## Experiment description and results
The goal of this experiment is to see how well we can learn a simple time series. 
There are two aspects to this:
1. How do we best featurize time series data?
2. Is the BOLT architecture sufficient to learn time series given appropriate features?

**Dataset:** synthetic store item demand forecasting dataset from Kaggle.
Train data is from 2013-01-01 to 2017-12-31
Test data is from 2018-01-01 to 2018-03-31 and does not have ground truth, so I split the train data into train and test sets.

First, I split the train data as follows:
* Train: 2013-01-01 to 2017-09-30
* Test: 2017-10-01 to 2017-12-31

3 setups:
* Dummy model: take the average for each store and item pair, and use that as the prediction for every day in the next 3 months -> 19.6% error (SMAPE)
* CMS count features + date features + BOLT with hidden dim 5000 and sparsity 0.05. More details in the "count feature configuration" subsection. -> 15.8% error (SMAPE) within 100 epochs
* Pandas features (windows + date features) + BOLT with hidden dim 10,000 and sparsity 0.05 -> 14.5% error (SMAPE)
Note: The count features used in the BOLT setup are lagged by at least 3 months since we are training BOLT to predict 3 months ahead.

They all perform worse than most solutions on Kaggle, which get around 12-13% error. But these scores are computed on the actual test data, which consists of records from 2018-01-01 to 2018-03-01.

Due to seasonalities in the data, training and testing with the following split should make our results more comparable to the results on Kaggle:
* Train: 2013-01-01 to 2016-12-31
* Test: 2017-01-01 to 2017-03-31

Made an error and split the data as follows:
* Train: 2013-01-01 to 2016-12-31
* Test: 2017-01-01 to 2017-09-30

Results:
* Dummy model got 26% error -> expected since it uses the same average to predict 9 months ahead
* CMS counts + date features + BOLT got 17% error -> while it also only uses records up to 2016-12-31 for training, it predicts with information from the test data. This is because the test data is over 9 months while the count feature lag is only 3 months.
* Pandas features + BOLT got 13.67% error. Same as above.

Corrected the data split and got the following results:
* Dummy model got 21% error
* CMS + date + BOLT got 19-20% error
* Pandas features + BOLT got 15.6% error after around 12 epochs. nan afterwards.

Tried increasing sparsity to 0.2 in the BOLT setups. Reduced the Pandas + BOLT setup error to 15%, but did not help the CMS + BOLT setup.

### Count feature configuration
All count features lag by at least 92 days since we predict 3 months ahead. 
Given a record `store_id,item_id,sales,timestamp`, the label for this record is `sales` and its features are some combination of windowed sums of `sales` from records with the same store_id and item_id but with a timestamp at least 92 days before `timestamp`.
* Daily sales from -92-7 to -92 days. Try to capture daily trends in the last 7 week of known sales data.
* Sum of sales over the week preceding -92-28, -92-21, -92-14, and -92-7 days. Try to capture weekly trend in the last month of known sales data.
* Sum of sales over the month preceding -92-28 and -92-58 days (the 2 months before -92-28). Try to captures monthly trends within last 3 months of known sales data.
* Sum of sales over the week preceding -92-30*n for n in 1 to 12; the amount sold for a week, around the same week, every previous month up to last 12 months / 1 year of known sales data. Try to capture monthly seasonality and sales around similar times of previous months, as well as sales around similar times of the previous year.

Other count windows I've tried that did not help:
* Monthly sales for each month in the last 12 months of known sales data. 

## Thoughts on results and next steps
* It's weird that restricting the test data to just the first 3 months of 2017 significantly increased the error of the BOLT setups, since BOLT was trained on the same portion of the dataset in both cases. Maybe the first quarter of 2017 is harder to predict compared to other quarters. 
    - [ ] check by training on the same 2013-2016 portion of the data but predict on the 2017-04-01 to 2017-09-30 portion, using 3-month lagged features. I expect to get much lower errors such that the high errors of the first quarter are balanced out.
* The Pandas + BOLT setup is relatively stable across prediction starting dates, so the BOLT engine itself seems fine
* The discrepancy between the Pandas and CMS BOLT setups can be attributed to several differences in the setup:
    * The pandas setup uses a much wider hidden layer. I could not run bolt with CMS with a wider hidden layer since I kept getting a “Killed” error. 
        - [ ] Seems like a memory leak. Fix it.
        - [ ] Also fix nan error
    * The pandas data pipeline adds small gaussian random noise to its counts, which helps prevent overfitting.
        - [ ] Remove from the pandas pipeline
    * The count min sketch can introduce unwanted noise through hash collisions. 
        - [x] Increase sketch width and height
        - [ ] Compare true counts against count min sketch counts 
    * The pandas pipeline uses different variants of moving average, e.g. exponentially weighted average, triangular windows, etc. 
        - [ ] Remove these variations from the pipeline and see how it affects performance 
* The CMS + BOLT setup performance is awfully close to the dummy model's performance. 
    - [x] Graph BOLT predictions in both CMS and Pandas setups against ground truth. Does it learn the overarching trend? Does it learn seasonalities? What does it fail to learn? In my opinion this is the most important next step.
* Splitting the training data like this is an unfair comparison with the Kaggle solutions because we have one less year's worth of training data. 
    - [ ] Instead of setting up our experiment to be comparable to the Kaggle leaderboard, we can run a Kaggle solution with our train/test split.
* Count features are taken from between 3 months before a record and 1 year 3 months before a record. This window may not be large enough.
    - [ ] Use count features up to 2 years before a record.
